KAFKA-8885: The Kafka Protocol should Support Optional Tagged Fields

cmccabe opened the pull request.

cmccabe committed:
KAFKA-8885: The Kafka Protocol should Support Optional Tagged Fields

twmb commented on clients/src/main/resources/common/message/README.md:
```diff
@@ -79,9 +87,14 @@ There are several primitive field types available.
+  In flexible versions, the length is a variable-length integer. 
 
 In addition to these primitive field types, there is also an array type.  Array
 types start with a "[]" and end with the name of the element type.  For
```
This file doesn't yet mention it, but I think it should include a section on arrays, especially so with this PR. Prior, all arrays are either non-nullable or nullable with the same length encoding as bytes. There is one special array with varint encoding for number of headers in a record (but that has a separate definition, I guess). With flexible versions, arrays are varint encoded.

twmb replied:
Also maybe a note in the nullable section that nullable strings, arrays, bytes are encoded with a length of -1.

cmccabe replied:
When I originally wrote this, I didn't intend this section as a complete description of how the encoding worked, but just a general hint to how much space various datatypes took, so that people could choose the right ones.  Maybe we should just include a link to the serialization documentation here, though, to avoid repeating that same information in two places and/or getting out of date.

ijuma replied:
@cmccabe Can you please share a link to the serialization documentation you mentioned?

cmccabe replied:
I was referring to https://kafka.apache.org/protocol.html .  There will be additional material there due to COMPACT_STRING, COMPACT_ARRAY, etc.  There should probably be something there about tagged fields as well.  Let's do that in a follow-on PR.

twmb commented on clients/src/main/resources/common/message/README.md:
```diff
@@ -79,9 +87,14 @@ There are several primitive field types available.
+* "string": a string.  Strings are serialized as a length followed by the
+  contents as UTF-8.  The contents must be less than 64kb in size.  In
+  non-flexible versions, the string length will always be 2 bytes.  In flexible
+  versions, the length is a variable-length integer. 
```
KIP-482 indicates that flexible versions will actually use unsigned variable length numbers, offset by 1 to reserve 0 indicating null. If that's the case, this should change from `variable-length integer`, which implies the old zig-zag varint, to `unsigned variable-length integer offset by 1, with 0 indicating null` (or something like that).

Also, it might be worth mentioning that the numbers are always offset by one, even if the field is non-nullable.

cmccabe replied:
I'm going to split the documentation part off into a separate PR, since there are some tricky questions about what should go in what documentation section.  Let's discuss it there.

cmccabe commented:
Hi all, I have split this PR up into several other PRs to be more reviewable, including #7372, #7344, #7340, and more to come.  I'm going to leave this one up for now so that people can see the bigger context of some of the changes, however.  Maybe I will rebase this one and use it for the last part once the other ones are in.  Thanks.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/ObjectSizeCache.java:
```diff
@@ -0,0 +1,54 @@
+ * It is intended to be used as part of a two-pass serialization process like:
+ * ObjectSizeCache cache = new ObjectSizeCache();
+ * message.calculateSize(version, cache);
+ * message.write(version, cache
```
Nit: `)` missing.

ijuma commented on clients/src/main/resources/common/message/CreateDelegationTokenResponse.json:
```diff
@@ -18,7 +18,10 @@
+  //
+  // Version 2 is the first flexible version.
+  "validVersions": "0-2",
+  "flexibleVersions": "2+",
```
How did you decide which requests should have flexible versions?

cmccabe replied:
I did the ones that are using auto-generation.

dajac commented on generator/src/main/java/org/apache/kafka/message/ApiMessageTypeGenerator.java:
```diff
@@ -245,6 +247,58 @@ private void generateToString() {
     }
 
+    private void generateHeaderVersion() {
+        buffer.printf("public short headerVersion(short version) {%n");
```
I wonder if it would make sense to create two different methods here to respectively retrieve the version of the request header and the version of the response header. I think we will need this to support the ApiVersions case. We could talk this in a separate PR or as part of KIP-511. What do you think?

cmccabe replied:
yes, let's do that.

dajac commented on clients/src/main/resources/common/message/ControlledShutdownResponse.json:
```diff
@@ -18,7 +18,10 @@
+  //
+  // Version 3 is the first flexible version.
+  "validVersions": "0-3",
+  "flexibleVersions": "3+",
```
Leaving the comment here because I haven't found a better place. It would be great if you could add tests in `RequestResponseTest#testSerialization` to cover all the versions which have been bumped.

cmccabe replied:
Hmm.  We don't generally do exhaustive testing of all versions in RequestResponseTest.  There would be a lot of entries!  But I agree we should add more stuff in RequestResponseTest.  Let's think about it in a follow-on PR

hachikuji replied:
+1 for additional tests. Testing every version is not crazy. We have been bitten a few times already due to untested version bumps.

dajac commented on generator/src/main/java/org/apache/kafka/message/ApiMessageTypeGenerator.java:
```diff
@@ -245,6 +247,58 @@ private void generateToString() {
+            short apiKey = entry.getKey();
+            buffer.printf("case %d:%n", apiKey);
+            buffer.incrementIndent();
+            if (apiKey == 7) {
```
I think this one deserve a small comment to explain the special case. It wasn't obvious to me as least.

cmccabe replied:
added

dajac reviewed (commented):
Thanks for the PR @cmccabe. Overall, it looks good to me. I have left few minor comments.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/ObjectSizeCache.java:
```diff
@@ -0,0 +1,54 @@
+ * message.calculateSize(version, cache);
+ * message.write(version, cache
+ */
+public final class ObjectSizeCache {
```
It looks like this map also stores byte arrays, so we should probably change the name and documentation?

cmccabe replied:
arrays are objects, right? :)
What names would be better?  SerializationSizeCache?  Just SizeCache?

ijuma replied:
No, I mean that we are storing the actual array as the value, not a size:

```java
public void cacheSerializedValue(Object o, byte[] val) {
    map.put(o, val);
}
```

cmccabe replied:
Good point.  I renamed this class to "ObjectSerializationCache" which should be clearer.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Readable.java:
```diff
@@ -26,34 +30,22 @@
+    default List<RawTaggedField> readRawTaggedField(List<RawTaggedField> unknowns, int tag, int size) {
+        if (unknowns == null) {
+            unknowns = new ArrayList<>();
         }
```
Are we doing this to avoid the allocation of `unknowns` unless there is at least one unknown?

cmccabe replied:
yes

ijuma commented on generator/src/test/java/org/apache/kafka/message/MessageDataGeneratorTest.java:
```diff
@@ -90,4 +90,192 @@ public void testInvalidNullDefaultForPotentiallyNonNullableArray() throws Except
     }
+
+    @Test
+    public void testInvalidFieldName() {
```
It would be helpful to indicate what is invalid about the name. Is it the underscore at the start?

cmccabe replied:
I added some JavaDoc

ijuma commented on generator/src/test/java/org/apache/kafka/message/MessageDataGeneratorTest.java:
```diff
@@ -90,4 +90,192 @@ public void testInvalidNullDefaultForPotentiallyNonNullableArray() throws Except
+                "    { \"name\": \"_badName\", \"type\": \"[]int32\", \"versions\": \"0+\" }",
+                "  ]",
+                "}")), MessageSpec.class);
+            fail("Expected MessageDataGenerator constructor to fail");
```
I would suggest using `assertThrows` in this and other tests added in this PR that validate that an exception is thrown.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/CompactArrayOf.java:
```diff
@@ -0,0 +1,139 @@
+ * A compact array represents its length with a varint rather than a
+ * fixed-length field.
+ */
+public class CompactArrayOf extends DocumentedType {
```
We also talked about arrays of primitives having a compact representation where tags are not needed per element. How would we describe such arrays, are they `packed arrays` versus `compact arrays`?

cmccabe replied:
They could be either ArrayOf and CompactArrayOf.  We don't have a separate array type for arrays of objects vs. arrays of non-object types

ijuma reviewed (commented):
Thanks for the PR. I took an initial pass and left some high level questions and a couple of nits.

cmccabe committed:
Revert message/README.md changes since we'll do those in a follow-on

cmccabe committed:
Fix merge conflicts

cmccabe committed:
split headerVersion into requestHeaderVersion and responseHeaderVersion

cmccabe committed:
address review comments

cmccabe commented:
* I have split the changes to message/README.md into a follow-on PR.

* I rebased on trunk and fixed a few minor conflicts.

* Split headerVersion into requestHeaderVersion and responseHeaderVersion.

ijuma commented:
There are some compiler errors after the latest updates.

cmccabe committed:
Fix some naming issues and awkward code

Rename ObjectSizeCache to ObjectSerializationCache

Prefix readable, writable, and size with an underscore in the generated
code to avoid conflicting with message fields that have these names.

Create MessageTestUtil.

cmccabe committed:
Use assertThrows in MessageDataGeneratorTest

cmccabe committed:
Fix compiler errors

cmccabe commented:
The compiler errors should be fixed now.

cmccabe committed:
Fix tagged array fields

* Fix code generation for tagged array fields

* Rename TestUUID to SimpleExampleMessage and add some tests for tagged fields there.

* Fix a bug in generating the code for tagged array fields

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/ObjectSerializationCache.java:
```diff
@@ -0,0 +1,57 @@
+
+    public int getArraySizeInBytes(Object o) {
+        Object value = map.get(o);
+        Integer sizeInBytes = (Integer) value;
```
Seems a bit messy to support different value types in the same map. Are we saving that much by not having separate maps?

cmccabe replied:
The memory overhead of having a separate map would be pretty large in the common case where objects are small.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/ObjectSerializationCache.java:
```diff
@@ -0,0 +1,57 @@
+        this.map = new IdentityHashMap<>();
+    }
+
+    public void setArraySizeInBytes(Object o, int size) {
```
I couldn't find any uses for this code in any of the generated classes. Do we have test cases which exercise this logic?

cmccabe replied:
The test message file SimpleExampleMessage.json contains a tagged array, which will use this logic.  I will add a test that uses that field.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Readable.java:
```diff
@@ -26,34 +30,22 @@
-        int length = readInt();
-        if (length < 0) {
-            return null;
+    default List<RawTaggedField> readRawTaggedField(List<RawTaggedField> unknowns, int tag, int size) {
```
Maybe `readUnknownTaggedField`?

hachikuji commented on clients/src/test/java/org/apache/kafka/common/protocol/types/RawTaggedFieldWriterTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Test
+    public void testWritingSeveralRawTaggedFields() {
+        List<RawTaggedField> tags = Arrays.asList(new RawTaggedField[] {
```
nit: use vararg constructor. A couple below as well

cmccabe replied:
ok

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Field.java:
```diff
@@ -115,6 +127,25 @@ public Array(String name, Type elementType, String docString) {
+        private static final String NAME = "_tagged_fields";
+        private static final String DOC_STRING = "The tagged fields";
+
+        public static TaggedFieldsSection of(Object... fields) {
```
nit: would be nice to document the expected type of `fields`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java:
```diff
@@ -0,0 +1,170 @@
+    @SuppressWarnings("unchecked")
+    public static TaggedFields of(Object... fields) {
+        TreeMap<Integer, Field> newFields = new TreeMap<>();
+        for (int i = 0; i < fields.length; i += 2) {
```
Maybe worth a check for even number of arguments.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Message.java:
```diff
@@ -37,25 +40,28 @@
+    int size(ObjectSerializationCache cache, short version);
 
     /**
      * Writes out this message to the given ByteBuffer.
```
Not from this patch, but should be `Writable` instead of `ByteBuffer`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/types/TaggedFields.java:
```diff
@@ -0,0 +1,170 @@
+            }
+            return objects;
+        } catch (ClassCastException e) {
+            throw new SchemaException("Not a Map.");
```
nit: not a `NavigableMap`

hachikuji commented on clients/src/main/resources/common/message/LeaderAndIsrRequest.json:
```diff
@@ -23,6 +23,7 @@
   //
   // Version 3 adds AddingReplicas and RemovingReplicas
```
Shall we bump this and `UpdateMetadata`?

cmccabe replied:
done

hachikuji commented on clients/src/main/resources/common/message/ControlledShutdownRequest.json:
```diff
@@ -24,7 +24,10 @@
   // Version 2 adds BrokerEpoch.
-  "validVersions": "0-2",
+  //
+  // Version 3 is the first flexible version.
```
Note that we will need a bump in `kafka.api.ApiVersion` for this. You will also have to update `KafkaServer` where we define the version to use statically. Perhaps we can just do this bump and the other inter-broker APIs in a separate patch.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:
```diff
@@ -794,49 +800,101 @@ private static void compareTypes(NamedType typeA, NamedType typeB) {
+            int size = message.size(cache, version);
+            ByteBuffer buf = ByteBuffer.allocate(size);
+            ByteBufferAccessor byteBufferAccessor = new ByteBufferAccessor(buf);
+            message.write(byteBufferAccessor, cache, version);
```
nit: you can use `assertThrows`. There are a few of these in this PR.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/message/MessageTest.java:
```diff
@@ -794,49 +800,101 @@ private static void compareTypes(NamedType typeA, NamedType typeB) {
+        while (alt.hasRemaining()) {
+            bld.append(String.format(" %02x", alt.get()));
+        }
+        System.out.println("WATERMELON: buffer is " + bld.toString());
```
Do we need this?

cmccabe replied:
removed

hachikuji commented on clients/src/test/resources/common/message/SimpleExampleMessage.json:
```diff
@@ -13,10 +13,13 @@
-}
\ No newline at end of file
+    { "name": "processId", "versions": "1+", "type": "uuid" },
+    { "name": "field1", "type": "[]int32", 
```
It would be useful to test a few more exotic variants, such as an optional array of structs.

cmccabe replied:
let's do this in a follow-on

hachikuji commented on generator/src/main/java/org/apache/kafka/message/ApiMessageTypeGenerator.java:
```diff
@@ -245,6 +248,59 @@ private void generateToString() {
+            short apiKey = entry.getKey();
+            buffer.printf("case %d:%n", apiKey);
+            buffer.incrementIndent();
+            if (apiKey == 7) {
```
Why don't we use `ApiKeys.CONTROLLED_SHUTDOWN.id()`?

cmccabe replied:
The generator module is built and run before the clients module where the ApiKeys class lives.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/message/ApiMessageTypeTest.java:
```diff
@@ -73,4 +73,25 @@ public void testUniqueness() {
+        assertEquals((short) 0, ApiMessageType.CREATE_TOPICS.responseHeaderVersion((short) 5));
+
+        assertEquals((short) 2, ApiMessageType.CREATE_TOPICS.requestHeaderVersion((short) 6));
+        assertEquals((short) 1, ApiMessageType.CREATE_TOPICS.responseHeaderVersion((short) 6));
```
These tests seem inconsistent with the generated code and fail when run locally. Seems like there's some confusion about whether the request and response headers bump at the same time. It might be simpler just to always bump them together, which is what the generated code appears to do.

cmccabe replied:
That works for every RPC except ApiVersions, which needs a v2 request header but a v0 response header.  It's definitely a wart, but hard to avoid when dealing with ApiVersionsRequest (see the KIP-511 discussions...)

hachikuji commented on generator/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -84,6 +100,56 @@ public FieldSpec(@JsonProperty("name") String name,
             }
         }
+        this.taggedVersions = Versions.parse(taggedVersions, Versions.NONE);
+        if ((flexibleVersions == null) || flexibleVersions.isEmpty()) {
```
nit: unneeded parenthesis

hachikuji commented on generator/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -84,6 +100,56 @@ public FieldSpec(@JsonProperty("name") String name,
+                    "override.");
+            }
+        }
+        this.tag = (tag == null) ? Optional.empty() : Optional.of(tag);
```
nit: `Optional.ofNullable(tag)`

hachikuji commented on generator/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -84,6 +100,56 @@ public FieldSpec(@JsonProperty("name") String name,
+        checkTagInvariants();
+    }
+
+    private void checkTagInvariants() {
```
Does it make sense to check flexible version invariants here as well?

cmccabe replied:
I think it's better to have them separate since the tags have their own invariants

hachikuji commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -281,6 +291,26 @@ private void generateFieldAccessors(StructSpec struct, boolean isSetElement) {
+        buffer.printf("if (_unknownTaggedFields == null) {%n");
+        buffer.incrementIndent();
+        headerGenerator.addImport(MessageGenerator.ARRAYLIST_CLASS);
+        buffer.printf("_unknownTaggedFields = new ArrayList<>(0);%n");
```
Could probably use `Collections.emptyList()` here

cmccabe replied:
the problem is that this field is mutable, and Collections.emptyList returns something immutable

hachikuji commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -624,133 +829,357 @@ private String readFieldFromStruct(FieldType type, String name) {
+            generate(buffer);
+    }
+
+    private void generateNonDefaultValueCheck(FieldSpec field) {
```
This is nifty

hachikuji commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -398,181 +418,365 @@ private void generateClassReader(String className, StructSpec struct,
+                        collectionType(arrayType.elementType().toString()), lengthVar);
+            } else {
+                headerGenerator.addImport(MessageGenerator.ARRAYLIST_CLASS);
+                buffer.printf("ArrayList<%s> newCollection = new ArrayList<%s>(%s);%n",
```
I wonder if you have given any thought to limiting allocations like this. For example, in the case of the byte array, we may be able to validate the size using the available bytes in the request

cmccabe replied:
I do think we're kind of goofy to allow arrays with 2**31 elements.  There must be a reasonable maximum we could set lower than that.  But there will probably be some compatibility implications to this, so it will take time to impose a reasonable limit now....

hachikuji commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -398,181 +418,365 @@ private void generateClassReader(String className, StructSpec struct,
+                    buffer.printf("this.%s = %s;%n", field.camelCaseName(), fieldDefault(field));
+                }).
+                ifMember(presentVersions -> {
+                    VersionConditional.forVersions(field.taggedVersions(), presentVersions).
```
Not a big deal, but there are a few cases where we could use a null check of `_taggedField` instead of a version check. Might make the generated code a little more readable.

cmccabe replied:
It would be a bit complex to change now since we're also filtering versions that aren't present at all and so on

hachikuji commented on generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java:
```diff
@@ -624,133 +829,357 @@ private String readFieldFromStruct(FieldType type, String name) {
+                            generateFieldToStruct(field, presentAndUntaggedVersions);
+                        }).
+                        ifMember(presentAndTaggedVersions -> {
+                            generateNonDefaultValueCheck(field);
```
I think there's a bug in the handling of nullable arrays when the default is not null. For example, consider the following field:

```json
    { "name": "field2", "type": "[]BlahType",
      "versions": "1+", "taggedVersions": "1+", "tag": 1,
      "nullableVersions": "1+",
      "fields": [
          { "name": "wootId", "versions": "1+", "type": "int32" },
      ]
    }
```
This results in the following code:

```java
        if (_version >= 1) {
            if (!field2.isEmpty()) {
                if (field2 == null) {
                    _taggedFields.put(1, null);
                } else {
                    Struct[] _nestedObjects = new Struct[field2.size()];
                    int i = 0;
                    for (BlahType element : this.field2) {
                        _nestedObjects[i++] = element.toStruct(_version);
                    }
                    _taggedFields.put(1, _nestedObjects);
                }
            }
        }
```
The null check should come first. Seems like the default value optimization needs to take into account nullable values. The same bug affects `size`.

In general, we probably need more testing, especially for default value handling.

cmccabe replied:
Thanks for finding this.  It might be better to address it in a follow on, since the fix could get complicated.  I'll push what I have for now.

hachikuji reviewed (commented):
Thanks, looking good overall. I left a few comments.

cmccabe committed:
Fix a bug where UUID objects were compared using == instead of equals()

cmccabe committed:
Add a test of tagged arrays to SimpleExampleMessageTest

cmccabe committed:
Address some review comments

cmccabe committed:
Do not use flexibleVersions for inter-broker RPCs (yet)

cmccabe committed:
Fix header stuff

cmccabe commented:
* Responded to all comments
* I rebased on trunk to catch up with the KIP-511 changes.
* Fixed the bugs in ApiMessageTypeTest and ApiMessageType.
* Fixed some other miscellaneous issues with ApiVersionsRequest.

I filed some follow-on JIRAs:
* KAFKA-8984: Improve tagged fields documentation
* KAFKA-8985: Use flexibleVersions with LeaderAndIsr, and improve RequestResponseTest coverage
* KAFKA-8986: Allow null as a valid default for tagged fields

cmccabe committed:
Fix SaslAuthenticatorTest

cmccabe committed:
Fix MessageTest

cmccabe committed:
Fix some unit tests and a networkclient bug

cmccabe committed:
Fix checkstyle

hachikuji commented:
Still at least one test failure. This one is reproducible locally:
```
12:22:00 kafka.server.ApiVersionsRequestTest > testApiVersionsRequest STARTED
12:22:06 kafka.server.ApiVersionsRequestTest.testApiVersionsRequest failed, log available in /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk11-scala2.13/core/build/reports/testOutput/kafka.server.ApiVersionsRequestTest.testApiVersionsRequest.test.stdout
12:22:06 
12:22:06 kafka.server.ApiVersionsRequestTest > testApiVersionsRequest FAILED
12:22:06     java.lang.OutOfMemoryError: Java heap space
12:22:06         at org.apache.kafka.common.utils.ImplicitLinkedHashCollection.clear(ImplicitLinkedHashCollection.java:566)
12:22:06         at org.apache.kafka.common.utils.ImplicitLinkedHashCollection.<init>(ImplicitLinkedHashCollection.java:530)
12:22:06         at org.apache.kafka.common.utils.ImplicitLinkedHashMultiCollection.<init>(ImplicitLinkedHashMultiCollection.java:52)
12:22:06         at org.apache.kafka.common.message.ApiVersionsResponseData$ApiVersionsResponseKeyCollection.<init>(ApiVersionsResponseData.java:615)
12:22:06         at org.apache.kafka.common.message.ApiVersionsResponseData.read(ApiVersionsResponseData.java:137)
12:22:06         at org.apache.kafka.common.message.ApiVersionsResponseData.<init>(ApiVersionsResponseData.java:87)
12:22:06         at org.apache.kafka.common.requests.ApiVersionsResponse.parse(ApiVersionsResponse.java:88)
12:22:06         at kafka.server.ApiVersionsRequestTest.sendApiVersionsRequest(ApiVersionsRequestTest.scala:81)
12:22:06         at kafka.server.ApiVersionsRequestTest.testApiVersionsRequest(ApiVersionsRequestTest.scala:48)
```

ijuma commented:
retest this please

ijuma committed:
Fix ApiVersionsRequestTest failure

ijuma commented:
Failures are all flakes.

ijuma commented:
retest this please

ijuma commented:
Tests passed locally:

> BUILD SUCCESSFUL in 42m 58s
> 150 actionable tasks: 139 executed, 11 up-to-date

hachikuji commented:
I will go ahead and merge. The failing tests are known to be flaky prior to this patch.

cmccabe commented:
Thanks, @ijuma and @hachikuji!  And all the other reviewers who helped with this

twmb commented:
Edit: concern retracted; after consideration, I think that tags on every struct level is fine.

cmccabe commented:
Hi @twmb, thanks for looking at this.  As you probably figured out (I see you edited your comment a bit), it's important to allow tagged fields to be added without a version bump.  Otherwise we don't get a lot of the benefits of a flexible schema.  This does require an extra byte per struct.  There was a lot of discussion about this on the mailing list.  The discussion period was actually much longer than the implementation period and definitely was not done at the last minute.  I looked for alternate solutions that didn't require the extra byte, but they were all very awkward and complex.

To counteract the extra space taken, we implemented more efficient serialization for strings, bytes, and arrays.  In the common case where these fields are small, we save between 1 and 3 bytes per object.  So if the objects in your hypothetical array of objects contain any of these things, the overhead is already cancelled out. 

I agree that it is annoying that an `[]int32` is now different from  `[]MySingleMemberStruct` which contains an int32 each.  The issue here is that there are a bunch of places where we pass around a short list of int32s to represent nodes, and it seemed excessive to add a byte of overhead to each.  Another issue is that there were optimizations for calculating the message size based on the fact that each entry was the same length.  I wanted to keep those optimizations.  So therefore, although it was a judgement call, I proposed a design where arrays of primitives could be serialized in the old, simpler way.

I hope this answers all the questions (and potential ones?) :) There is more discussion about this on the mailing list if you want to go in depth.  I always appreciate feedback and I made a point of pulling in some Kafka client authors before this was finalized.
